pandas/core/arrays/string_.py

```diff
@@ -731,6 +733,22 @@ def _putmask(self, mask: npt.NDArray[np.bool_], value) -> None:
         # base class implementation that uses __setitem__
         ExtensionArray._putmask(self, mask, value)
 
+    def isin(self, values: ArrayLike) -> npt.NDArray[np.bool_]:
+        if not isinstance(values, BaseStringArray):
+            if not lib.is_string_array(np.asarray(values), skipna=True):
```

Comment:
Is it worth short-circuiting for the case where `values` is all `NA` and `self._hasna` is `False`?

Reply:
Also, is it worth short-circuiting this check when the `values.dtype` is `ArrowDtype(pa.string())`?

Reply:
> Is it worth short-circuiting for the case where `values` is all `NA` and `self._hasna` is `False`?

I would say that is probably not worth the added complexity for something that is just speeding up an uncommon corner case? (also, checking `self._hasna` has a cost)

> Also, is it worth short-circuiting this check when the `values.dtype` is `ArrowDtype(pa.string())`?

I wanted to do that, but then I would need to import ArrowDtype and pyarrow inline (because this file is for the object string dtype, that works without pyarrow), so at that point I was again unsure if we should add a special case here

(I think a typical use case is passing a list, and that will at the moment never be coerced to ArrowDtype before it gets here)

Now, what we should maybe do to essentially speed up this case is to ensure `lib.is_string_array` shortcuts on pyarrow strings and directly returns True, and ensure that `StringArray._from_sequence` shortcuts on pyarrow strings to ensure the most efficient conversion (if it doesn't do that already).

Reply:
Although I can probably check for `ArrowDtype(pa.string())` without importing it by comparing for equality with `"string[pyarrow]"`